location/lbs/activity/audioset/dataprocessing/audio_processing.py

```diff
@@ -0,0 +1,577 @@
+    removed from the audio_dict dictionary.
+
+    Args:
+        dest_dir: Path to the parent directory where the downloaded videos are
+            to be stored.
+        audio_dict: A dictionary with video_id, AudioSetEntry key, value pairs
+        redo: A boolean of specifying whether to re-download all the YouTube
+            videos.
+    """
+    failed_download_set = get_failed_downloads(dest_dir)
+    failed_downloads = []
+    for video_id, entry in audio_dict.items():
+        if video_id in failed_download_set:
+            failed_downloads.append(video_id)
+            continue
+        success = download(dest_dir, video_id, failed_downloads, redo)
+        start_time = entry.start_time
+        end_time = entry.end_time
+        if success:
+            chop_audio(dest_dir, video_id, start_time, end_time)
+    for video_id in failed_downloads:
+        del audio_dict[video_id]
+
+
+def download(dest_dir, video_id, failed_downloads, redo):
+    """Downloads a YouTube video using its video_id
+
+    Calls the youtube-dl tool to download a YouTube video by its video_id and
+    convert the video into an audio file. If the download is successful,
+    it stores the video_id in a list, downloaded_video.
+
+    Args:
+        dest_dir: Path to the parent directory where the downloaded videos are
+            to be stored.
+        video_id: the video_id of the YouTube video to be downloaded.
+        failed_downloads: a list where failed downloaded YouTube videos'
+            video_id will be stored.
+        redo: A boolean of specifying whether to re-download all the YouTube
+            videos.
+
+    Returns:
+        A boolean of whether the download was successful or not. Returns false
+        if the video was already downloaded and redo is False.
+    """
+    check_dir(dest_dir)
+    # check to see if the video_id has already been downloaded and skip the
+    # download if it is already downloaded unless the -r, --redo flag has been
+    # passed
+    video_path = join(dest_dir, 'yt_videos', 'sliced_' + video_id + '.wav')
+    already_downloaded = isfile(video_path)
+    if already_downloaded and not redo:
+        logging.info('Already Downloaded')
+        return False
+
+    ydl_opts = {
+        'format': 'bestaudio/best',
+        'postprocessors': [{
+            'key': 'FFmpegExtractAudio',
+            'preferredquality': '192',
+        }],
+        # Force the file naming of outputs.
+        'outtmpl': join(dest_dir, 'tmp', video_id + '.%(ext)s')
+    }
+    with youtube_dl.YoutubeDL(ydl_opts) as ydl:
+        try:
+            ydl.download(['https://www.youtube.com/watch?v=' + video_id])
+            logging.info('Download Complete')
+            return True
+        except youtube_dl.DownloadError:
+            store_failed_download(dest_dir, video_id)
+            failed_downloads.append(video_id)
+            logging.info('Downloading Failed.')
```

Comment:
In addition to logging that download filed, it'll be helpful to log the id of the video that failed to download. I would consider logging at the end of the script where the failed downloads were logged to.

Reply:
I do have a running list of video_ids that failed to download stored in a text file. For every failed download, its video_id is stored in the failed_downloads.txt file in the destination directory, via the store_failed_download(dest_dir, video_id) function. Should I log the video_ids in addition to doing this?